pxemgr/ipxe_handlers.go

```diff
@@ -349,129 +289,110 @@ func (mgr *pxeManagerT) setState(serial string, w http.ResponseWriter, r *http.R
+	assignedIPs := map[string]struct{}{}
+	for _, host := range mgr.cluster.GetAllHosts() {
+		assignedIPs[host.InternalAddr.String()] = struct{}{}
 	}
 
-	err = host.Commit("updated host overrides: " + strings.Join(updatedVars, ", "))
-	if err != nil {
-		w.WriteHeader(500)
-		w.Write([]byte("committing updated host overrides failed"))
-		return
+	IPisAvailable := func(ip net.IP) bool {
+		_, exists := assignedIPs[ip.String()]
+		return !exists
 	}
-	mgr.cluster.Update()
-	w.WriteHeader(202)
-}
 
-func (mgr *pxeManagerT) setEtcdClusterToken(serial string, w http.ResponseWriter, r *http.Request) {
-	host, exists := mgr.cluster.HostWithSerial(serial)
-	if !exists {
-		w.WriteHeader(400)
-		w.Write([]byte("host doesn't exist"))
-		return
-	}
+	currentIP := net.ParseIP(mgr.config.Network.PrimaryNIC.IPRange.Start)
+	rangeEnd := net.ParseIP(mgr.config.Network.PrimaryNIC.IPRange.End)
 
-	decoder := json.NewDecoder(r.Body)
-	payload := hostmgr.Host{}
-	err := decoder.Decode(&payload)
-	if err != nil {
-		w.WriteHeader(400)
-		w.Write([]byte("unable to parse json data in set_etcd_cluster_token request"))
-		return
+	for ; ; ipLessThanOrEqual(currentIP, rangeEnd) {
+		if IPisAvailable(currentIP) {
+			return currentIP
+		}
+		currentIP = incIP(currentIP)
 	}
+}
 
-	host.EtcdClusterToken = payload.EtcdClusterToken
-	err = host.Commit("updated host etcd cluster token")
-	if err != nil {
-		w.WriteHeader(500)
-		w.Write([]byte("committing updated host etcd cluster token failed"))
-		return
+func (mgr *pxeManagerT) getNextAdditionalIP(nicIndex int, nicName string) net.IP {
+	assignedIPs := map[string]struct{}{}
+	for _, host := range mgr.cluster.GetAllHosts() {
+		if _, exists := host.AdditionalAddrs[nicName]; exists {
+			assignedIPs[host.AdditionalAddrs[nicName].String()] = struct{}{}
+		}
 	}
-	mgr.cluster.Update()
-	w.WriteHeader(202)
-}
 
-func (mgr *pxeManagerT) welcomeHandler(w http.ResponseWriter, r *http.Request) {
-	w.WriteHeader(200)
-	w.Write([]byte("this is the iPXE server of mayu " + mgr.version))
-	return
-}
+	IPisAvailable := func(ip net.IP) bool {
+		_, exists := assignedIPs[ip.String()]
+		return !exists
+	}
 
-func (mgr *pxeManagerT) hostCoreOSVersion(r *http.Request) string {
-	coreOSversion := mgr.config.DefaultCoreOSVersion
+	currentIP := net.ParseIP(mgr.config.Network.ExtraNICs[nicIndex].IPRange.Start)
+	rangeEnd := net.ParseIP(mgr.config.Network.ExtraNICs[nicIndex].IPRange.End)
 
-	host, exists := mgr.hostFromSerialVar(r)
-	if exists {
-		if version, exist := host.Overrides["CoreOSVersion"]; exist {
-			return version.(string)
+	for ; ; ipLessThanOrEqual(currentIP, rangeEnd) {
+		if IPisAvailable(currentIP) {
+			return currentIP
 		}
+		currentIP = incIP(currentIP)
+	}
+}
 
-		if host.CoreOSVersion == "" {
-			return mgr.config.DefaultCoreOSVersion
-		} else {
-			return host.CoreOSVersion
+// check af all hosts have properly assigned IP addresses to all Network.ExtraNICs entries
+func (mgr *pxeManagerT) checkAdditionalNICAddresses() {
```

Comment:
this is an important part of code which on boot checks if proper IPs are assigned to all Extra NICs as there can be a scenario when customer want to add new NICs to already installed cluster or the range could change